File: 2025-03-29/Add_Two_Numbers.java
package amazon;

public class Add_Two_Numbers {
    public static class ListNode {
      int val;
      ListNode next;
      ListNode() {}
      ListNode(int val) { this.val = val; }
      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode ans = new ListNode();
        ListNode temp = ans;
        int carry = 0;

        while(l1 != null || l2 != null || carry > 0) {
            int d1 = l1 != null ? l1.val : 0;
            int d2 = l2 != null ? l2.val : 0;

            int sum = d1 + d2 + carry;
            ans.next = new ListNode(sum % 10);
            carry = sum / 10;

            l1 = l1 != null ? l1.next : null;
            l2 = l2 != null ? l2.next : null;
            ans = ans.next;
        }

        return temp.next;
    }
    public static void main(String[] args) {
        ListNode l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
        ListNode l2 = new ListNode(5, new ListNode(6, new ListNode(4)));

        addTwoNumbers(l1, l2);
    }
}

File: 2026-02-02/Add_Two_Numbers.py
from typing import Optional
from Linked_List import Linked_List, ListNode


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        result = ListNode(0)
        cur = result
        carry = 0

        while l1 or l2 or carry:
            sum = carry
            if l1:
                sum += l1.val
                l1 = l1.next
            if l2:
                sum += l2.val
                l2 = l2.next

            carry = sum // 10
            cur.next = ListNode(sum % 10)
            cur = cur.next

        return result.next


sol = Solution()
ll = Linked_List()
head1 = ll.build_from_list([2,4,3])
head2 = ll.build_from_list([5,6,4])
ll.print_list(sol.addTwoNumbers(head1, head2))